interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,384 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package requestprompts
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces/prompting"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify/listener"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	ErrNotFound = errors.New("cannot find prompt with the given ID for the given user")
+)
+
+// Prompt contains information about a request for which a user should be
+// prompted.
+type Prompt struct {
+	ID           string             `json:"id"`
+	Timestamp    time.Time          `json:"timestamp"`
+	Snap         string             `json:"snap"`
+	Interface    string             `json:"interface"`
+	Constraints  *PromptConstraints `json:"constraints"`
+	listenerReqs []*listener.Request
+}
+
+// PromptConstraints are like prompting.Constraints, but have a "path" field
+// instead of a "path-pattern", and include the available permissions for the
+// interface corresponding to the prompt.
+type PromptConstraints struct {
+	Path                 string   `json:"path"`
+	Permissions          []string `json:"permissions"`
+	AvailablePermissions []string `json:"available-permissions"`
+}
+
+// equals returns true if the two prompt constraints are identical.
+func (pc *PromptConstraints) equals(other *PromptConstraints) bool {
+	if pc.Path != other.Path || len(pc.Permissions) != len(other.Permissions) {
+		return false
+	}
+	// Avoid using reflect.DeepEquals to compare []string contents
+	for i := range pc.Permissions {
+		if pc.Permissions[i] != other.Permissions[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// subtractPermissions removes all of the given permissions from the list of
+// permissions in the constraints.
+func (pc *PromptConstraints) subtractPermissions(permissions []string) (modified bool) {
+	newPermissions := make([]string, 0, len(pc.Permissions))
+	for _, perm := range pc.Permissions {
```

Comment:
This is quadratic.

Reply:
for context afair the lists here are very short, otoh I don't remember if we can make sure they are sorted which could helps a slightly bit using SortedListContains instead

Reply:
They are ordered according to the order of available permissions for the given interface. E.g. `["read", "write", "execute"]`, `["read", "execute"]`, etc. So we can't use `SortedListContains` here. But the number of permissions should be at most ~5 (read, write, execute, append, lock) for any interface I can think of, I believe. And currently only 3 (rwx).